File: app/src/main/java/com/radio/app/models/AppSettings.java
package com.radio.app.models;

import android.content.Context;
import android.content.SharedPreferences;

import org.json.JSONArray;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class AppSettings {
    private static final String PREFS_NAME = "radio_app_settings";
    private static AppSettings instance;

    public static synchronized AppSettings getInstance(Context context) {
        if (instance == null) {
            instance = new AppSettings();
            instance.load(context);
        }
        return instance;
    }

    private void load(Context context) {
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        aiModel = prefs.getString("ai_model", AI_MODEL_WENXIN);
        asrProvider = prefs.getString("asr_provider", ASR_BAIDU);
        uiTheme = prefs.getString("ui_theme", THEME_DARK);
        preloadCache = prefs.getBoolean("preload_cache", false);
        preloadCacheCount = prefs.getInt("preload_cache_count", 1);
        enablePreprocessing = prefs.getBoolean("enable_preprocessing", false);
        preprocessingCount = prefs.getInt("preprocessing_count", 1);
        audioFocus = prefs.getBoolean("audio_focus", true);
        continuousPlay = prefs.getBoolean("continuous_play", true);
        String dislikedJson = prefs.getString("disliked_episodes", "[]");
        try {
            JSONArray arr = new JSONArray(dislikedJson);
            dislikedEpisodes.clear();
            for (int i = 0; i < arr.length(); i++) {
                dislikedEpisodes.add(arr.getString(i));
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void save(Context context) {
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = prefs.edit();
        editor.putString("ai_model", aiModel);
        editor.putString("asr_provider", asrProvider);
        editor.putString("ui_theme", uiTheme);
        editor.putBoolean("preload_cache", preloadCache);
        editor.putInt("preload_cache_count", preloadCacheCount);
        editor.putBoolean("enable_preprocessing", enablePreprocessing);
        editor.putInt("preprocessing_count", preprocessingCount);
        editor.putBoolean("audio_focus", audioFocus);
        editor.putBoolean("continuous_play", continuousPlay);
        JSONArray arr = new JSONArray();
        for (String id : dislikedEpisodes) arr.put(id);
        editor.putString("disliked_episodes", arr.toString());
        editor.apply();
    }

    public void addDislikedEpisode(Context context, String episodeId) {
        if (!dislikedEpisodes.contains(episodeId)) {
            dislikedEpisodes.add(episodeId);
            save(context);
        }
    }

    public void removeDislikedEpisode(Context context, String episodeId) {
        dislikedEpisodes.remove(episodeId);
        save(context);
    }

    public boolean isDisliked(String episodeId) {
        return dislikedEpisodes.contains(episodeId);
    }

    public static final String AI_MODEL_WENXIN = "wenxin";
    public static final String AI_MODEL_DEEPSEEK = "deepseek";
    public static final String AI_MODEL_QWEN = "qwen";
    public static final String AI_MODEL_FUNASR = "funasr";
    public static final String AI_MODEL_WHISPER = "whisper";
    public static final String AI_MODEL_JIU_AI_TING = "jiu-ai-ting";

    public static final String ASR_BAIDU = "baidu";
    public static final String ASR_FUNASR = "funasr";
    public static final String ASR_WHISPER = "whisper";
    public static final String ASR_VOSK = "vosk";

    public static final String THEME_DARK = "dark";
    public static final String THEME_FRESH = "fresh";
    public static final String THEME_CLASSIC = "classic";
    public static final String THEME_MINIMAL = "minimal";
    public static final String THEME_CUSTOM = "custom";

    public static final String SUBTITLE_SMALL = "small";
    public static final String SUBTITLE_MEDIUM = "medium";
    public static final String SUBTITLE_LARGE = "large";
    public static final String LANG_CN = "zh-CN";
    public static final String LANG_EN = "en-US";

    private String aiModel = AI_MODEL_WENXIN;
    private boolean autoSkipWater = true;
    private boolean autoDownload = false;
    private boolean autoCache = false;
    private String subtitleSize = SUBTITLE_MEDIUM;
    private String subtitleLanguage = LANG_CN;
    private String voiceLanguage = LANG_CN;
    private String asrProvider = ASR_BAIDU;
    private String uiTheme = THEME_DARK;
    private CustomColors customColors = new CustomColors();
    private KeywordConfig keywordConfig = new KeywordConfig();
    private boolean preloadCache = false;
    private int preloadCacheCount = 1;
    private boolean enablePreprocessing = false;
    private int preprocessingCount = 1;
    private boolean audioFocus = true;
    private boolean continuousPlay = true;
    private List<String> dislikedEpisodes = new ArrayList<>();

    public String getAiModel() { return aiModel; }
    public void setAiModel(String aiModel) { this.aiModel = aiModel; }
    public String getAsrProvider() { return asrProvider; }
    public void setAsrProvider(String asrProvider) { this.asrProvider = asrProvider; }
    public String getUiTheme() { return uiTheme; }
    public void setUiTheme(String uiTheme) { this.uiTheme = uiTheme; }
    public CustomColors getCustomColors() { return customColors; }
    public void setCustomColors(CustomColors customColors) { this.customColors = customColors; }
    public KeywordConfig getKeywordConfig() { return keywordConfig; }
    public void setKeywordConfig(KeywordConfig keywordConfig) { this.keywordConfig = keywordConfig; }
    public boolean isPreloadCache() { return preloadCache; }
    public void setPreloadCache(boolean preloadCache) { this.preloadCache = preloadCache; }
    public int getPreloadCacheCount() { return preloadCacheCount; }
    public void setPreloadCacheCount(int preloadCacheCount) { this.preloadCacheCount = preloadCacheCount; }
    public boolean isEnablePreprocessing() { return enablePreprocessing; }
    public void setEnablePreprocessing(boolean enablePreprocessing) { this.enablePreprocessing = enablePreprocessing; }
    public int getPreprocessingCount() { return preprocessingCount; }
    public void setPreprocessingCount(int preprocessingCount) { this.preprocessingCount = preprocessingCount; }
    public boolean isAudioFocus() { return audioFocus; }
    public void setAudioFocus(boolean audioFocus) { this.audioFocus = audioFocus; }
    public boolean isContinuousPlay() { return continuousPlay; }
    public void setContinuousPlay(boolean continuousPlay) { this.continuousPlay = continuousPlay; }
    public boolean isAutoSkipWater() { return autoSkipWater; }
    public void setAutoSkipWater(boolean autoSkipWater) { this.autoSkipWater = autoSkipWater; }
    public boolean isAutoDownload() { return autoDownload; }
    public void setAutoDownload(boolean autoDownload) { this.autoDownload = autoDownload; }
    public boolean isAutoCache() { return autoCache; }
    public void setAutoCache(boolean autoCache) { this.autoCache = autoCache; }
    public String getSubtitleSize() { return subtitleSize; }
    public void setSubtitleSize(String subtitleSize) { this.subtitleSize = subtitleSize; }
    public String getSubtitleLanguage() { return subtitleLanguage; }
    public void setSubtitleLanguage(String subtitleLanguage) { this.subtitleLanguage = subtitleLanguage; }
    public String getVoiceLanguage() { return voiceLanguage; }
    public void setVoiceLanguage(String voiceLanguage) { this.voiceLanguage = voiceLanguage; }
    public List<String> getDislikedEpisodes() { return dislikedEpisodes; }
    public void setDislikedEpisodes(List<String> dislikedEpisodes) { this.dislikedEpisodes = dislikedEpisodes; }

    public static class CustomColors {
        private String primary = "#0f3460";
        private String accent = "#e94560";
        private String background = "#1a1a2e";
        private String text = "#eaeaea";
        private String card = "#16213e";
        private String border = "#0f3460";
        private String success = "#4ecca3";
        private String warning = "#f4d03f";

        public String getPrimary() { return primary; } public void setPrimary(String primary) { this.primary = primary; }
        public String getAccent() { return accent; } public void setAccent(String accent) { this.accent = accent; }
        public String getBackground() { return background; } public void setBackground(String background) { this.background = background; }
        public String getText() { return text; } public void setText(String text) { this.text = text; }
        public String getCard() { return card; } public void setCard(String card) { this.card = card; }
        public String getBorder() { return border; } public void setBorder(String border) { this.border = border; }
        public String getSuccess() { return success; } public void setSuccess(String success) { this.success = success; }
        public String getWarning() { return warning; } public void setWarning(String warning) { this.warning = warning; }
    }

    public static class KeywordConfig {
        private List<String> dryKeywords = new ArrayList<>(Arrays.asList("新闻", "资讯", "报道", "访谈", "评论"));
        private List<String> waterKeywords = new ArrayList<>(Arrays.asList("广告", "音乐", "歌曲", "休息"));
        private List<String> contentDryKeywords = new ArrayList<>();
        private List<String> contentWaterKeywords = new ArrayList<>();
        private String dryLogic = "or";
        private String waterLogic = "or";
        private String contentDryLogic = "or";
        private String contentWaterLogic = "or";

        public List<String> getDryKeywords() { return dryKeywords; } public void setDryKeywords(List<String> dryKeywords) { this.dryKeywords = dryKeywords; }
        public List<String> getWaterKeywords() { return waterKeywords; } public void setWaterKeywords(List<String> waterKeywords) { this.waterKeywords = waterKeywords; }
        public List<String> getContentDryKeywords() { return contentDryKeywords; } public void setContentDryKeywords(List<String> contentDryKeywords) { this.contentDryKeywords = contentDryKeywords; }
        public List<String> getContentWaterKeywords() { return contentWaterKeywords; } public void setContentWaterKeywords(List<String> contentWaterKeywords) { this.contentWaterKeywords = contentWaterKeywords; }
        public String getDryLogic() { return dryLogic; } public void setDryLogic(String dryLogic) { this.dryLogic = dryLogic; }
        public String getWaterLogic() { return waterLogic; } public void setWaterLogic(String waterLogic) { this.waterLogic = waterLogic; }
        public String getContentDryLogic() { return contentDryLogic; } public void setContentDryLogic(String contentDryLogic) { this.contentDryLogic = contentDryLogic; }
        public String getContentWaterLogic() { return contentWaterLogic; } public void setContentWaterLogic(String contentWaterLogic) { this.contentWaterLogic = contentWaterLogic; }
    }
}

File: app/src/main/java/com/radio/app/models/AppSettings.kt
package com.radio.app.models

import android.content.Context
import android.content.SharedPreferences
import org.json.JSONArray

class AppSettings private constructor() {

    companion object {
        private const val PREFS_NAME = "radio_app_settings"

        const val AI_MODEL_WENXIN = "wenxin"
        const val AI_MODEL_DEEPSEEK = "deepseek"
        const val AI_MODEL_QWEN = "qwen"
        const val AI_MODEL_FUNASR = "funasr"
        const val AI_MODEL_WHISPER = "whisper"
        const val AI_MODEL_JIU_AI_TING = "jiu-ai-ting"
        const val AI_MODEL_MNN_LLM = "mnn-llm"

        const val ASR_BAIDU = "baidu"
        const val ASR_FUNASR = "funasr"
        const val ASR_WHISPER = "whisper"
        const val ASR_VOSK = "vosk"

        const val THEME_DARK = "dark"
        const val THEME_FRESH = "fresh"
        const val THEME_CLASSIC = "classic"
        const val THEME_MINIMAL = "minimal"
        const val THEME_CUSTOM = "custom"

        const val SUBTITLE_SMALL = "small"
        const val SUBTITLE_MEDIUM = "medium"
        const val SUBTITLE_LARGE = "large"
        const val LANG_CN = "zh-CN"
        const val LANG_EN = "en-US"

        const val SPLIT_MODE_NONE = "none"
        const val SPLIT_MODE_HORIZONTAL = "horizontal"
        const val SPLIT_MODE_VERTICAL = "vertical"

        @Volatile
        private var instance: AppSettings? = null

        fun getInstance(context: Context): AppSettings {
            return instance ?: synchronized(this) {
                instance ?: AppSettings().also {
                    it.load(context)
                    instance = it
                }
            }
        }
    }

    var aiModel: String = AI_MODEL_WENXIN
    var asrProvider: String = ASR_BAIDU
    var voskModelDir: String = ""  // [v2.0.61] Issue 6 Fix: Save selected Vosk model directory name
    var whisperModelDir: String = ""  // [v2.2.9] Save selected Whisper model directory name (whisper-tiny, whisper-base, etc.)
    // [v2.0.50] Issue 3 Fix: Track Whisper crash count to permanently disable after 2 crashes
    var whisperCrashCount: Int = 0
    var forceVoskUntil: Long = 0  // [v2.2.8] Force Vosk after Whisper crashes (timestamp until which to skip Whisper)
    var uiTheme: String = THEME_DARK
    var subtitleSize: String = SUBTITLE_MEDIUM
    var subtitleLanguage: String = LANG_CN
    var voiceLanguage: String = LANG_CN
    var splitMode: String = SPLIT_MODE_NONE
    var customColors: CustomColors = CustomColors()
    var keywordConfig: KeywordConfig = KeywordConfig()
    var preloadCache: Boolean = false
    var preloadCacheCount: Int = 10
    var enablePreprocessing: Boolean = true
    var preprocessingCount: Int = 1
    var audioFocus: Boolean = true
    var continuousPlay: Boolean = true
    var autoSkipWater: Boolean = true
    var autoDownload: Boolean = false
    var autoCache: Boolean = false
    var savePlaybackPosition: Boolean = true
    var rememberLastEpisode: Boolean = true
    var wifiOnlyPreCache: Boolean = true
    var notificationStyle: String = "compact"
    var skipSeconds: Int = 15
    var dislikedEpisodes: MutableList<String> = mutableListOf()
    var stationPlayCount: MutableMap<String, Int> = mutableMapOf()
    var lastSelectedDate: String = ""
    var lastSelectedStationId: String = ""
    var pinduoduoDetectionInterval: Int = 5  // [v2.1.2] seconds, default 5

    /** Gson 安全访问：确保字段不为 null */
    fun safeSubtitleSize(): String = subtitleSize ?: SUBTITLE_MEDIUM
    fun safeSubtitleLanguage(): String = subtitleLanguage ?: LANG_CN
    fun safeVoiceLanguage(): String = voiceLanguage ?: LANG_CN
    fun safeUiTheme(): String = uiTheme ?: THEME_DARK
    fun safeAiModel(): String = aiModel ?: AI_MODEL_WENXIN
    fun safeAsrProvider(): String = asrProvider ?: ASR_BAIDU
    fun safeSplitMode(): String = splitMode ?: SPLIT_MODE_NONE

    /**
     * v2.0.97 - Reload ASR provider and Vosk model dir from SharedPreferences.
     * SubtitleGeneratorService runs in a separate process (:subtitle), whose AppSettings
     * singleton is loaded once at process start and never refreshed. When user changes ASR
     * engine in UI process, the subtitle process's singleton is stale. This method re-reads
     * the latest values from SharedPreferences.
     */
    fun reloadAsrSettings(context: Context) {
        // [v2.4.0] Force reload from disk — SharedPreferences doesn't sync across processes
        // (UI process writes, :subtitle process has stale cache). The broadcast handler writes
        // to this process's prefs, but if the broadcast wasn't received, we need to force-read.
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_MULTI_PROCESS)
        prefs.all  // Force access to trigger reload
        asrProvider = prefs.getString("asr_provider", ASR_BAIDU) ?: ASR_BAIDU
        voskModelDir = prefs.getString("vosk_model_dir", "") ?: ""
        whisperModelDir = prefs.getString("whisper_model_dir", "") ?: ""  // [v2.2.9]
        // [v2.1.2] Also reload pinduoduo detection interval for RadioPlaybackService
        pinduoduoDetectionInterval = prefs.getInt("pinduoduo_detection_interval", 5)
    }

    private fun load(context: Context) {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_MULTI_PROCESS)
        aiModel = prefs.getString("ai_model", AI_MODEL_WENXIN) ?: AI_MODEL_WENXIN
        asrProvider = prefs.getString("asr_provider", ASR_BAIDU) ?: ASR_BAIDU
        voskModelDir = prefs.getString("vosk_model_dir", "") ?: ""  // [v2.0.61] Issue 6
        whisperModelDir = prefs.getString("whisper_model_dir", "") ?: ""  // [v2.2.9]
        whisperCrashCount = prefs.getInt("whisper_crash_count", 0)  // [v2.0.50]
        forceVoskUntil = prefs.getLong("force_vosk_until", 0)  // [v2.2.8]
        uiTheme = prefs.getString("ui_theme", THEME_DARK) ?: THEME_DARK
        preloadCache = prefs.getBoolean("preload_cache", false)
        preloadCacheCount = prefs.getInt("preload_cache_count", 1)
        autoCache = prefs.getBoolean("auto_cache", false)
        enablePreprocessing = prefs.getBoolean("enable_preprocessing", true)
        preprocessingCount = prefs.getInt("preprocessing_count", 1)
        audioFocus = prefs.getBoolean("audio_focus", true)
        continuousPlay = prefs.getBoolean("continuous_play", true)
        savePlaybackPosition = prefs.getBoolean("save_playback_position", true)
        rememberLastEpisode = prefs.getBoolean("remember_last_episode", true)
        wifiOnlyPreCache = prefs.getBoolean("wifi_only_precache", true)
        notificationStyle = prefs.getString("notification_style", "compact") ?: "compact"
        skipSeconds = prefs.getInt("skip_seconds", 15)
        splitMode = prefs.getString("split_mode", SPLIT_MODE_NONE) ?: SPLIT_MODE_NONE

        val dislikedJson = prefs.getString("disliked_episodes", "[]") ?: "[]"
        try {
            val arr = JSONArray(dislikedJson)
            dislikedEpisodes.clear()
            for (i in 0 until arr.length()) {
                dislikedEpisodes.add(arr.getString(i))
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

        // 加载上次选择的日期和电台
        lastSelectedDate = prefs.getString("last_selected_date", "") ?: ""
        lastSelectedStationId = prefs.getString("last_selected_station_id", "") ?: ""
        pinduoduoDetectionInterval = prefs.getInt("pinduoduo_detection_interval", 5)  // [v2.1.2]

        // 加载播放次数
        val playCountJson = prefs.getString("station_play_count", "{}") ?: "{}"
        try {
            val obj = org.json.JSONObject(playCountJson)
            stationPlayCount.clear()
            val keys = obj.keys()
            while (keys.hasNext()) {
                val key = keys.next()
                stationPlayCount[key] = obj.getInt(key)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun save(context: Context) {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_MULTI_PROCESS)
        prefs.edit().apply {
            putString("ai_model", aiModel)
            putString("asr_provider", asrProvider)
            putString("vosk_model_dir", voskModelDir)  // [v2.0.61] Issue 6
            putString("whisper_model_dir", whisperModelDir)  // [v2.2.9]
            putInt("whisper_crash_count", whisperCrashCount)  // [v2.0.50]
            putLong("force_vosk_until", forceVoskUntil)  // [v2.2.8]
            putString("ui_theme", uiTheme)
            putBoolean("preload_cache", preloadCache)
            putInt("preload_cache_count", preloadCacheCount)
            putBoolean("auto_cache", autoCache)
            putBoolean("enable_preprocessing", enablePreprocessing)
            putInt("preprocessing_count", preprocessingCount)
            putBoolean("audio_focus", audioFocus)
            putBoolean("continuous_play", continuousPlay)
            putBoolean("save_playback_position", savePlaybackPosition)
            putBoolean("remember_last_episode", rememberLastEpisode)
            putBoolean("wifi_only_precache", wifiOnlyPreCache)
            putString("notification_style", notificationStyle)
            putInt("skip_seconds", skipSeconds)
            putString("split_mode", splitMode)
            val arr = JSONArray()
            dislikedEpisodes.forEach { arr.put(it) }
            putString("disliked_episodes", arr.toString())
            val playCountObj = org.json.JSONObject()
            stationPlayCount.forEach { (k, v) -> playCountObj.put(k, v) }
            putString("station_play_count", playCountObj.toString())
            putString("last_selected_date", lastSelectedDate)
            putString("last_selected_station_id", lastSelectedStationId)
            putInt("pinduoduo_detection_interval", pinduoduoDetectionInterval)  // [v2.1.2]
            apply()
        }
    }

    fun addDislikedEpisode(context: Context, episodeId: String, stationId: String? = null, title: String? = null) {
        var changed = false
        if (!dislikedEpisodes.contains(episodeId)) {
            dislikedEpisodes.add(episodeId)
            changed = true
        }
        // 同时存储基于 title 的 key，使 dislike 过滤能跨天匹配同一节目
        if (!stationId.isNullOrBlank() && !title.isNullOrBlank()) {
            val titleKey = "$stationId::$title"
            if (!dislikedEpisodes.contains(titleKey)) {
                dislikedEpisodes.add(titleKey)
                changed = true
            }
        }
        if (changed) save(context)
    }

    fun removeDislikedEpisode(context: Context, episodeId: String) {
        dislikedEpisodes.remove(episodeId)
        save(context)
    }

    fun isDisliked(episodeId: String): Boolean = dislikedEpisodes.contains(episodeId)

    /**
     * 按节目名称判断是否不喜欢（每天这个节目都不喜欢）
     * 兼容《》括号： disliked列表可能有《标题》，但实际标题可能没有《》
     */
    fun isDislikedByTitle(stationId: String?, title: String?): Boolean {
        if (stationId.isNullOrBlank() || title.isNullOrBlank()) return false
        val key = "$stationId::$title"
        if (dislikedEpisodes.contains(key)) return true
        // 尝试去除《》括号后匹配
        val strippedTitle = title.replace(Regex("^《|》$"), "")
        if (strippedTitle != title) {
            val strippedKey = "$stationId::$strippedTitle"
            if (dislikedEpisodes.contains(strippedKey)) return true
        }
        // 尝试添加《》括号后匹配
        val wrappedTitle = "《$title》"
        val wrappedKey = "$stationId::$wrappedTitle"
        if (dislikedEpisodes.contains(wrappedKey)) return true
        // 归一化标题：去除所有标点符号后比较（兼容 ·、-、（）等标点差异）
        // 说明：v2.0.39 的子串模糊匹配过于激进，会把用户喜欢的节目误判为不喜欢。
        // 现改为精确的标点归一化匹配——仅当去掉所有标点后标题完全相同（或仅差「重播」后缀）时才命中。
        val normalizeTitle = { s: String -> s.replace(Regex("[《》·\\-—（）()\\[】【]"), "") }
        val normTitle = normalizeTitle(title)
        val normStrippedTitle = normalizeTitle(strippedTitle)
        for (dislikedKey in dislikedEpisodes) {
            if (!dislikedKey.startsWith("$stationId::")) continue
            val dislikedTitle = dislikedKey.removePrefix("$stationId::")
            val strippedDisliked = dislikedTitle.replace(Regex("^《|》$"), "")
            val normDisliked = normalizeTitle(strippedDisliked)
            if (normDisliked == normTitle || normDisliked == normStrippedTitle) return true
            // 兼容（重播）后缀：disliked 标题去除标点后加上"重播"与当前标题匹配
            if (normDisliked + "重播" == normTitle || normDisliked + "重播" == normStrippedTitle) return true
        }
        return false
    }

    fun toggleDislikedEpisode(context: Context, episode: com.radio.app.models.Episode): Boolean {
        val stationId = episode.stationId ?: return false
        val title = episode.title ?: return false
        val key = "$stationId::$title"
        val epId = episode.id
        // 判断是否已 dislike：title-based key 或 episode ID 任一存在即视为已 dislike
        val isCurrentlyDisliked = dislikedEpisodes.contains(key) || dislikedEpisodes.contains(epId)
        return if (isCurrentlyDisliked) {
            // --- 取消 dislike：移除所有相关 key ---
            // 移除 title-based key 及其变体（有/无《》括号）
            dislikedEpisodes.remove(key)
            val strippedTitle = title.replace(Regex("^《|》$"), "")
            if (strippedTitle != title) {
                val strippedKey = "$stationId::$strippedTitle"
                dislikedEpisodes.remove(strippedKey)
            }
            val wrappedTitle = "《$title》"
            val wrappedKey = "$stationId::$wrappedTitle"
            dislikedEpisodes.remove(wrappedKey)
            // 移除所有模糊匹配的 title-based key
            val toRemove = mutableListOf<String>()
            for (dislikedKey in dislikedEpisodes) {
                if (!dislikedKey.startsWith("$stationId::")) continue
                val dislikedTitle = dislikedKey.removePrefix("$stationId::")
                val strippedDisliked = dislikedTitle.replace(Regex("^《|》$"), "")
                if (strippedDisliked == title || strippedDisliked == strippedTitle) {
                    toRemove.add(dislikedKey)
                }
            }
            dislikedEpisodes.removeAll(toRemove)
            // 同时移除 episode ID（精确匹配）
            dislikedEpisodes.remove(epId)
            save(context)
            false
        } else {
            // --- 设置 dislike：同时存储 title-based key 和 episode ID ---
            dislikedEpisodes.add(key)       // title-based key，跨天匹配同一节目
            dislikedEpisodes.add(epId)      // episode ID，精确匹配当天那一集
            save(context)
            // 同时保存到 all_episodes 持久化存储，确保 dislike 筛选能跨天找到该节目
            if (!episode.audioUrl.isNullOrBlank()) {
                try {
                    val prefs = context.getSharedPreferences("all_episodes", android.content.Context.MODE_PRIVATE)
                    prefs.edit().putString(episode.audioUrl, com.google.gson.Gson().toJson(episode)).apply()
                } catch (e: Exception) { /* ignore */ }
            }
            true
        }
    }

    fun incrementStationPlayCount(context: Context, stationId: String) {
        val count = stationPlayCount.getOrDefault(stationId, 0) + 1
        stationPlayCount[stationId] = count
        save(context)
    }

    fun getStationPlayCount(stationId: String): Int {
        return stationPlayCount.getOrDefault(stationId, 0)
    }

    data class CustomColors(
        var primary: String = "#0f3460",
        var accent: String = "#e94560",
        var background: String = "#1a1a2e",
        var text: String = "#eaeaea",
        var card: String = "#16213e",
        var border: String = "#0f3460",
        var success: String = "#4ecca3",
        var warning: String = "#f4d03f"
    )

    data class KeywordConfig(
        var dryKeywords: MutableList<String> = mutableListOf("新闻", "资讯", "报道", "访谈", "评论"),
        var waterKeywords: MutableList<String> = mutableListOf("广告", "音乐", "歌曲", "休息"),
        var contentDryKeywords: MutableList<String> = mutableListOf(),
        var contentWaterKeywords: MutableList<String> = mutableListOf(),
        var dryLogic: String = "or",
        var waterLogic: String = "or",
        var contentDryLogic: String = "or",
        var contentWaterLogic: String = "or"
    )
}
